sdk/compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Parsing.hs

```diff
@@ -163,18 +193,66 @@ parseClientMessageWithTracker tracker val = pickReqMethodTo tracker $ \extract -
       (Right (LSP.FromClientRsp (Pair method (Const mHome)) rsp), Just newIxMap)
     Left msg -> (Left msg, Nothing)
 
--- Takes a message from server and stores it if its a request, so that later messages from the client can deduce response context
-putServerReq :: MethodTrackerVar 'LSP.FromServer -> FilePath -> LSP.FromServerMessage -> IO ()
-putServerReq tracker home msg =
-  case msg of
-    LSP.FromServerMess meth mess ->
-      case LSP.splitServerMethod meth of
-        LSP.IsServerReq ->
-          let LSP.RequestMessage {_id, _method} = mess
-            in putReqMethodSingleFromServer tracker home _id _method
-        LSP.IsServerEither ->
-          case mess of
-            LSP.ReqMess LSP.RequestMessage {_id, _method} -> putReqMethodSingleFromServer tracker home _id _method
-            _ -> pure ()
-        _ -> pure ()
-    _ -> pure ()
+-- Map.mapAccum where the replacement value is a Maybe. Accumulator is still updated for `Nothing` values
+mapMaybeAccum :: Ord k => (a -> b -> (a, Maybe c)) -> a -> Map.Map k b -> (a, Map.Map k c)
+mapMaybeAccum f z = flip Map.foldrWithKey (z, Map.empty) $ \k v (accum, m) ->
+  second (maybe m (\v' -> Map.insert k v' m)) $ f accum v
+
+-- Convenience for the longwinded FromClient Some TrackedMethod type
+type SomeFromClientTrackedMethod = Some @(LSP.Method 'LSP.FromClient 'LSP.Request) TrackedMethod
+
+{-# ANN adjustClientTrackers ("HLint: ignore Avoid restricted function" :: String) #-}
```

Comment:
Which restricted function are we avoiding here?

Reply:
Added a comment, its the spooky `unsafeCoerce`, needed for phantom variable coercion